File: 2020/6.c
#include <stdio.h>
#include <stdlib.h>
#include "aoc.h"

static int parse(const char *str, int *x, int *y)
{
    int c, n, *p;

    hinit();
    n = *x = *y = 0;

    printf("%s\n", str);
    while (*str != '\0') {
        /* each person's answers are delimited by newlines */
        if (*str != '\n') {
            c = *str;
            p = hintlookup(c);
            if (p == NULL) {
                hintinsert(c, 1);
                (*x)++;
            } else
                (*p)++;
        } else
            n++;
        str++;
    }

    for (c = 'a'; c <= 'z'; c++) {
        p = hintlookup(c);
        if (p != NULL)
            *y += (*p == n) ? 1 : 0;
    }
    printf("x = %d, y = %d\n", *x, *y);

    hfree(free);
    return n;
}

int main(void)
{
    int i, c, p, na, nb, x, y;
    char group[1024];

    na = nb = p = i = 0;
    do {
        c = getchar();
        /* group entries are delimited by blank lines */
        if ((c == '\n' && p == '\n') || c == EOF) {
            group[i] = '\0';
            parse(group, &x, &y);
            na += x;
            nb += y;
            i = 0;
        } else {
            group[i] = c;
            i++;
        }
        p = c;
    } while (i < 1023 && c != EOF);

    printf("A) the sum of counts is %d\n", na);
    printf("B) the sum of counts is %d\n", nb);

    return 0;
}

File: 2020/zig/src/6.zig
const std = @import("std");

fn parse(allocator: *std.mem.Allocator, str: []u8, na: *u32, nb: *u32) !void {
    var map = std.AutoHashMap(u32, u32).init(allocator);
    defer map.deinit();

    var answers = std.mem.tokenize(str, "\n");
    var n: u32 = 0;
    while (answers.next()) |ans| {
        for (ans) |c| {
            if (map.get(c)) |p| {
                try map.put(c, p + 1);
            } else {
                try map.put(c, 1);
                na.* += 1;
            }
        }
        n += 1;
    }

    var c: u8 = 'a';
    while (c <= 'z') : (c += 1) {
        if (map.get(c)) |p| {
            if (p == n) {
                nb.* += 1;
            }
        }
    }
}

pub fn main() anyerror!void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    var p: u8 = undefined;
    var na: u32 = 0;
    var nb: u32 = 0;
    var group = std.ArrayList(u8).init(allocator);
    defer group.deinit();

    var stdin = std.io.getStdIn().reader();
    var in = std.io.bufferedReader(stdin).reader();
    while (in.readByte() catch null) |c| {
        // Group entries delimited by blank lines
        if (c == '\n' and p == '\n') {
            try parse(allocator, group.items, &na, &nb);
            group.shrinkRetainingCapacity(0);
        } else {
            try group.append(c);
        }
        p = c;
    }
    try parse(allocator, group.items, &na, &nb); // final group before EOF

    std.debug.print("A) the sum of counts is {d}\n", .{na});
    std.debug.print("B) the sum of counts is {d}\n", .{nb});
}
